### PR TITLE
[Feature] Implement Batch Gradient Descent

### PR DESCRIPTION
**Pull request by WallNutss**
Implementing batch gradient descent for updating the weights and bias itself. Previously,  even thought that I have split the dataset into mini batches, which is for an example if the dataset is 100, and the batch size is 10, that means there will be 10 mini batches in every mini batches is 100, but as you can see in the [code](https://github.com/WallNutss/neural-network-csharp/blob/2b1b5952951645983b1c59909047ae33d7d8fe49/NeuralNetworkCSharp/Core/Network.cs#L606), even though I have split them into mini batches, I still input them one by one and immediately update the weights and bias. Making it not really correct as this prone to bias as every weight & bias update are dependent on each data sample.

So, I think rather than that, what if like for an example get the batch size set to like 5, the different is that rather update the parameters in every sample, we will get the $\nabla_{W}$ and $\nabla_{b}$, and average them based on the batch size and then we can update the weight and bias. So, it should be inline with the source that if the loss function is define like this

$\nabla_{L} \approx \frac{1}{n} \sum_{x} C_x$ 

Where this is tell us that the value of the loss gradient is equal to the sum of the average of the gradient of the parameters. We can extend them to this branch thingy, when updating the network parameters, we can average the each of the gradients $\nabla_{L}$ of the current gradients parameter of the data and by aggregate them first along the batch data that as you can see on the `AccumulatedGradients` function. Then average them when updating the weights and biases in this formula, which is in the end achieving the average of the loss function, WITH... achieve it in every weight and biases parameter after all the batch data has been processed

$W_{k}^{'} = W_{k} - \frac{eta}{m} \nabla{W_k}$ 

From the formula above, by this also do the same thing that we previously do, which is if we want to still input them one by one and immediately update the weight and biases without averaging them, we can make the batch size into `1`, so it will only average one single point data, as the batch will contain only one data, making it in the end the same as when we input them one by one when updating it.

### CODE REVIEW

**Comment by WallNutss** [comment]
I will merge this for now, from my current knowledge, what I do now is right. Comment for the bad result of the model which is using batch gradient method, when I try set the batch size equal to 1, I got more or less the same result as the previous architecture. Turn's out that with current batch gradient method, if I also set the learning rate same as for an example 0.01 and the batch size is 32, the multiplier for the gradient descent step will like equal to $\frac{0.01}{32} \approx 0.0003125$, where this is the root cause why the model perform bad, as the step learning rate is to small to achieve meaningful learning to get parameter to represent the model
Screenshot 2025-01-18 194939
